[TEST-FAILURE] ClientCacheEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq

Failed here: https://github.com/hazelcast/hazelcast/pull/16961#issuecomment-625114772


The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/572/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/572/Hazelcast-pr-builder-572.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientCacheEventJournalBasicTest>AbstractEventJournalBasicTest.allowReadingWithFutureSeq:512->HazelcastTestSupport.assertOpenEventually:1108->HazelcastTestSupport.assertOpenEventually:1123 CountDownLatch failed to complete within 30 seconds, count left: 1
[INFO] 
[ERROR] Tests run: 27737, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

--- a/hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalBasicTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/journal/AbstractEventJournalBasicTest.java
@@ -32,7 +32,6 @@ import com.hazelcast.spi.properties.ClusterProperty;
 import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastTestSupport;
 import org.junit.Before;
-import org.junit.Ignore;
 import org.junit.Test;
 
 import java.util.ArrayList;
@@ -49,6 +48,7 @@ import java.util.function.BiConsumer;
 import java.util.function.Function;
 import java.util.function.Predicate;
 
+import static com.hazelcast.internal.util.ExceptionUtil.rethrow;
 import static com.hazelcast.internal.util.MapUtil.createHashMap;
 import static com.hazelcast.journal.EventJournalEventAdapter.EventType.ADDED;
 import static com.hazelcast.journal.EventJournalEventAdapter.EventType.EVICTED;
@@ -476,7 +476,6 @@ public abstract class AbstractEventJournalBasicTest<EJ_TYPE> extends HazelcastTe
     }
 
     @Test
-    @Ignore("https://github.com/hazelcast/hazelcast/issues/16964")
     public void allowReadingWithFutureSeq() throws Exception {
         final EventJournalTestContext<String, Integer, EJ_TYPE> context = createContext();
 
@@ -485,34 +484,34 @@ public abstract class AbstractEventJournalBasicTest<EJ_TYPE> extends HazelcastTe
         assertEquals(-1, state.getNewestSequence());
         assertEventJournalSize(context.dataAdapter, 0);
 
-        final String key = randomPartitionKey();
         final Integer value = RANDOM.nextInt();
         final CountDownLatch latch = new CountDownLatch(1);
         final int startSequence = 1;
 
         final BiConsumer<ReadResultSet<EJ_TYPE>, Throwable> callback = (response, t) -> {
             if (t == null) {
+                latch.countDown();
                 assertEquals(1, response.size());
                 final EventJournalEventAdapter<String, Integer, EJ_TYPE> journalAdapter = context.eventJournalAdapter;
                 final EJ_TYPE e = response.get(0);
 
                 assertEquals(ADDED, journalAdapter.getType(e));
-                assertEquals(key, journalAdapter.getKey(e));
                 assertEquals(value, journalAdapter.getNewValue(e));
-                assertNotEquals(startSequence + response.readCount(), response.getNextSequenceToReadFrom());
-                assertEquals(1, response.getNextSequenceToReadFrom());
-                latch.countDown();
             } else {
-                t.printStackTrace();
+                rethrow(t);
             }
         };
-        readFromEventJournal(context.dataAdapter, startSequence, 1, partitionId, TRUE_PREDICATE, IDENTITY_FUNCTION)
-                .whenCompleteAsync(callback);
+        CompletionStage<ReadResultSet<EJ_TYPE>> callbackStage =
+                readFromEventJournal(context.dataAdapter, startSequence, 1, partitionId, TRUE_PREDICATE, IDENTITY_FUNCTION)
+                    .whenCompleteAsync(callback);
 
-        context.dataAdapter.put(key, value);
+        assertTrueEventually(() -> {
+            context.dataAdapter.put(randomPartitionKey(), value);
+            assertTrue(latch.await(200, TimeUnit.MILLISECONDS));
+        }, 30);
 
-        assertOpenEventually(latch, 30);
-        assertEventJournalSize(context.dataAdapter, 1);
+        // ensure no exception thrown from callback
+        callbackStage.toCompletableFuture().join();
     }
 
     /**

